executor/src/witgen/jit/processor.rs

```diff
@@ -411,48 +465,94 @@ impl<'a, T: FieldElement> Processor<'a, T> {
-                let solved = identities
-                    .into_iter()
-                    .map(
-                        |(expression, row_offset)| match witgen.evaluate(expression, row_offset) {
+                let is_solved = pairs
+                    .iter()
+                    .map(move |(expression, row_offset)| {
+                        match witgen.evaluate(expression, *row_offset) {
                             None => false,
                             Some(value) => value.try_to_known().is_some(),
-                        },
-                    )
+                        }
+                    })
                     .collect_vec();
 
-                let solved_count = solved.iter().filter(|v| **v).count();
-                let unsolved_prefix = solved.iter().take_while(|v| !**v).count();
-                let unsolved_suffix = solved.iter().rev().take_while(|v| !**v).count();
+                let solved_count = is_solved.iter().filter(|v| **v).count();
+                let unsolved_prefix = is_solved.iter().take_while(|v| !**v).count();
+                let unsolved_suffix = is_solved.iter().rev().take_while(|v| !**v).count();
                 // There need to be at least `self.block_size` solved identities
                 // and there can be unsolved rows at the start or at the end, but not
                 // in the middle.
-                solved_count >= self.block_size
-                    && solved_count + unsolved_prefix + unsolved_suffix == solved.len()
+                if solved_count >= self.block_size
+                    && solved_count + unsolved_prefix + unsolved_suffix == is_solved.len()
+                {
+                    vec![]
+                } else {
+                    pairs
+                }
+                .into_iter()
+                .map(move |(_, row_offset)| (identity, row_offset))
             })
     }
+
+    /// Try to set the given variables to the first value in their allowed range,
+    /// as long as this does not create a conflict.
+    fn guess_unknown_variables<FixedEval: FixedEvaluator<T>>(
+        &self,
+        unknown_variables: &[Variable],
+        can_process: impl CanProcessCall<T>,
+        witgen: &mut WitgenInference<'a, T, FixedEval>,
+        mut identity_queue: IdentityQueue<'a, '_, T>,
+    ) -> bool {
+        let mut tentative_witgen = witgen.clone();
+        // TODO: We could also call `process_until_no_progress` after each variable
+        // and revert if this caused an error and skip those variables.
+        // It might be that the variables will be determined by other variables.
+        // An example is `XIsZero` and `XInv`.
+        for var in unknown_variables {
+            let value = tentative_witgen.range_constraint(var).range().0;
+            match tentative_witgen.set_variable(var.clone(), value) {
+                Err(_) => return false,
+                Ok(updated_vars) => identity_queue.variables_updated(updated_vars),
```

Comment:
Do we want to mutate the passed identity queue here? I think we want to clone it in the beginning of the function and only mutate that?

Reply:
Good point! We should assert that it is empty at the start and empty at the end.

Reply:
Ah wait, we receive a fresh copy, it's `mut identity_queue: IdentityQueue<'a, '_, T>,`, not `identity_queue: &mut IdentityQueue<'a, '_, T>,